impl/client/src/main/javascript/web/util/_a11y.js

```diff
@@ -35,6 +35,358 @@ define(["./_focus"], function(focusUtil) {
+     *
+     * @param evt - The keydown event object
+     */
+    function onKeyPress(evt) {
+      var key = evt.which || evt.keyCode;
+      var lastActiveId = activeDescendantId;
+      activeDescendantId = elem.getAttribute("aria-activedescendant");
+      var allOptions = getAllOptions();
+      var currentItem = getActiveOption() || allOptions[0];
+      var nextItem = currentItem;
+
+      if (!currentItem) {
+        return;
+      }
+
+      switch (key) {
+        case keyCodes.pageUp:
+        case keyCodes.pageDown:
+          break;
+        case keyCodes.arrowUp:
+        case keyCodes.arrowDown:
+          evt.preventDefault();
+          if (!activeDescendantId) {
+            // focus first option if no option was previously focused, and perform no other actions
+            setActiveOption(currentItem);
+            break;
+          }
+
+          if (key === keyCodes.arrowUp) {
+            nextItem = findPreviousOption(currentItem);
+          } else {
+            nextItem = findNextOption(currentItem);
+          }
+
+          if (nextItem && event.shiftKey) {
+            selectRange(startRangeIndex, nextItem);
+          }
+
+          if (nextItem) {
+            setActiveOption(nextItem);
+            evt.preventDefault();
+          }
+
+          break;
+        case keyCodes.home:
+          evt.preventDefault();
+          focusFirstItem();
+
+          if (evt.shiftKey && evt.ctrlKey) {
+            selectRange(startRangeIndex, 0);
+          }
+          break;
+        case keyCodes.end:
+          evt.preventDefault();
+          focusLastItem();
+
+          if (evt.shiftKey && evt.ctrlKey) {
+            selectRange(startRangeIndex, allOptions.length - 1);
+          }
+          break;
+        case keyCodes.shift:
+          startRangeIndex = getOptionIndex(currentItem);
+          break;
+        case keyCodes.space:
+          evt.preventDefault();
+          toggleSelectItem(nextItem);
+          break;
+        case 65:
+          // handle control + A
+          if (evt.ctrlKey || evt.metaKey) {
+            evt.preventDefault();
+            selectRange(0, allOptions.length - 1);
+            break;
+          }
+          // fall through for Type-ahead
+        default:
+          var character = String.fromCharCode(key).toUpperCase();
+          var startIndex = !keysSoFar ? getActiveOptionIndex() : null;
+
+          keysSoFar += character;
+
+          clearKeysSoFarAfterDelay();
+
+          var optionToFocus = findOption(keysSoFar, startIndex);
+          if (optionToFocus) {
+            setActiveOption(optionToFocus);
+          }
+          break;
+      }
+
+      if (activeDescendantId !== lastActiveId) {
+        updateScroll();
+      }
+    }
+
+    /**
+     * Helper function to check if a number is within a range; no side effects.
+     *
+     * @param index
+     * @param start
+     * @param end
+     * @returns {boolean}
```

Comment:
Missing parameter data types.
We use `@return` preferably to `@returns`.
Missing description text.